### PR TITLE
Allow setting selection

### PR DESCRIPTION
**Pull request by mikehaverstock**
This adds the ability to set the mathquill selection. We already had the ability to read it. In principle there was enough information in the the `{latex, startIndex, and endIndex}` to be able to restore the selection but in practice it's much more useful to have a path to a node and then a number of times to either simulate a `shift-left` or `shift-right`. There doesn't appear to be a way in mathquill to simple set the selection between two nodes. The way you must do it is based on finding a node and then growing the selection in a direction. Saving the instructions to find the the anchor and then optionally growing the selection with a certain number of left or rights makes setting the selection pretty straightforward.

Because the way we are indexing into the MQ tree is with a series of `Down` and `Right` commands it's pretty critical that we have the exact same tree structure. This means it's not really possible / practical to generate a desired selection programatically. The intended usecase is for you to get a selection out of mathquill and maybe sometime later attempt to restore it. We put the critical info for restoring it within a `opaqueSnapshot` property on the selection. This gives us room to at a later time amend / change the underlying data if we want. There's a latex cleaning step that happens as part of the selection generation that complicates things. It's one of the reasons we didn't try to just use the `startIndex` and `endIndex` we already have. We store the original uncleanedLatex to compare against because even though unnecessary spaces aren't important to displaying the latex they are critical for the tree structure we will index into.

## Tests
- [x] double check that all of our previous tests that we can `get` the selection also work to `set` the selection
- [x] add explicit tests for single cursor on a complicated equation. Test every position is settable.
- [x] double check edge cases:
  - [x] empty selection can be set
  - [x] entire selection can be set
  - [x] selection is not changed if the `originalLatex` does not match.
 - [x] selection encodes and restores anchor and head

### CODE REVIEW

File: src/mathquill.d.ts [hunk 1]
```diff
@@ -19,6 +19,17 @@ declare namespace MathQuill {
...
     };
 
+    type ExportedLatexSelection = {
+      latex: string;
+      startIndex: number;
+      endIndex: number;
+      opaqueSnapshot: {
+        uncleanedLatex: string;
+        cursorInsertPath: string;
+        selectionLength: number;
```
Comment (anandthakker) [inline src/mathquill.d.ts:29]: Do we care about making this actually opaque from the point of view of consumers?  We could do something like:

```ts
type OpaqueSelectionSnapshot = { readonly __selectionSnapshot: unique symbol }

type InternalSelectionSnapshot = {
    uncleanedLatex: string;
    cursorInsertPath: string;
    selectionLength: number;
}

// return type and input type for `selection()`
type ExportedLatexSelection = {
  latex: string;
  startIndex: number;
  endIndex: number;
  // within mathquill we just cast this to InternalSelectionSnapshot
  opaqueSnapshot: OpaqueSelectionSnapshot
}
```

File: src/services/latex.ts [hunk 2]
```diff
@@ -116,18 +127,132 @@ class Controller_latex extends Controller_keystroke {
...
+
+  private insertCursorAtPath(path: string): boolean {
+    if (!path) return false;
+
+    let node: MQNode = this.root;
+
+    // we generate the path starting from the node working up to the root.
+    // So we need to work backwards when following the path. The very last
+    // instruction we encounter is not a reference to a node. It's a reference
+    // to where within the node we should insert. Either RightOf the node or
```
Comment (anandthakker) [inline src/services/latex.ts:156]:
```suggestion
    // So we need to work backwards when following the path. The
    // very last instruction we encounter do not point to a node. They point to a space where a cursor could be inserted: either just to the right of the current node ("R") or just to the left of the current node's children ("D")
```

Comment (anandthakker) [reply src/services/latex.ts:156]: ^ suggested edit now that we're not using `RightOf` and `LeftEnd`

File: src/services/latex.ts [hunk 3]
```diff
@@ -116,18 +127,132 @@ class Controller_latex extends Controller_keystroke {
...
+          selectDir(dir);
+        }
+      });
+    }
+  }
+
+  // any time there's a selection there is a cursor and anticursor. The
+  // anticursor is the anchor, and the cursor is the head. It should be
+  // true that these are siblings. If you trace right or left far enough
+  // you will reach the other other. This returns the direction and magnitude
```
Comment (anandthakker) [inline src/services/latex.ts:213]:
```suggestion
  // you will reach the other one. This returns the direction and magnitude
```

File: src/services/latex.ts [hunk 4]
```diff
@@ -116,18 +127,132 @@ class Controller_latex extends Controller_keystroke {
...
+  // of how many hops it takes to find the cursor from the anticursor. Otherwise
+  // returns 0. The idea is to try this both with L and R and use the one, if any,
+  // that comes back with a non-zero answer.
```
Comment (anandthakker) [inline src/services/latex.ts:216]: Nice comment

File: src/services/latex.ts [hunk 5]
```diff
@@ -116,18 +127,132 @@ class Controller_latex extends Controller_keystroke {
...
+
+  exportLatexSelection(): ExportedLatexSelection {
     var ctx: LatexContext = {
       latex: '',
       startIndex: -1,
       endIndex: -1
     };
 
+    let cursorInsertPath: string = '';
+    let selectionLength: number = 0;
```
Comment (anandthakker) [inline src/services/latex.ts:242]: Since this function is also finding the selected latex, "length" might be a little confusing. Suggested rename: `signedSelectionSize`?  And the function could be `findSignedSelectionSize`.  (I get that "size" is just as ambiguous as "length", just trying to find a different word to use)

File: test/unit/latex.test.js [hunk 6]
```diff
@@ -370,6 +671,20 @@ suite('latex', function () {
...
+        const originalSnapShot = mq.selection();
+        mq2.latex('');
+        mq2.latex(originalSnapShot.opaqueSnapshot.uncleanedLatex);
+        mq2.selection(originalSnapShot);
+        const restoredSnapShot = mq2.selection();
+        assert.equal(
+          JSON.stringify(restoredSnapShot, null, 2),
+          JSON.stringify(originalSnapShot, null, 2),
+          'can restore selection'
+        );
```
Comment (anandthakker) [inline test/unit/latex.test.js:687]: Nice

**Review by anandthakker — APPROVED** [review]
Logic and tests all look great.  Comments are all just naming / documentation stuff, which I was paying extra attention to because of how MQ is its own, often-confusing world

File: src/mathquill.d.ts (hunk 1 above)
Comment (mikehaverstock) [reply src/mathquill.d.ts]: I'm not sure if this should be completely opaque or not. What about in knox tests where we want to assert an assertion matches what we expect? I could imagine hardcoding in a few opaqueSnapshot dumps into our assertions and types not being happy. I think I'm personally fine just naming it opaqueSnapshot for now. Maybe after actually using it in knox I'll have a stronger opinion one way or the other.

**Review by anandthakker — APPROVED** [review]
